composer/datasets/c4.py

```diff
@@ -0,0 +1,333 @@
+        # Samples are read into and randomly sampled from per-device shuffle buffer
+        if self.shuffle:
+            text_dataset = text_dataset.shuffle(buffer_size=self.shuffle_buffer_size, seed=self.seed)
+
+        # Map text samples to token samples
+        # NOTE: Mapping is executed in batched mode for better CPU utilization,
+        # but the returned dataset is still an iterable over tokenized samples
+        text_sample_batch_size = 1000
+        token_dataset = text_dataset.map(
+            self._tokenize,
+            batched=True,
+            batch_size=text_sample_batch_size,
+        )
+
+        if self.group_method != "truncate":
+            # Map variable-length token samples to fixed-length token samples
+            # NOTE: Mapping is executed in batched mode for better CPU utilization,
+            # but the returned dataset is still an iterable over tokenized samples.
+            # NOTE: Depending on the 'group_method', this step may alter the number of
+            # token samples in the dataset, and may mix neighboring token samples together.
+            token_sample_batch_size = 1000
+            token_dataset = token_dataset.map(
+                self._group_tokens,
+                batched=True,
+                batch_size=token_sample_batch_size,
+            )
+
+        # Limit the number of post-processed samples
+        self.iterable_dataset = token_dataset.take(self.max_samples_per_device)
+
+    def __iter__(self):
+        worker_info = get_worker_info()
+        if worker_info is not None and worker_info.num_workers != 1:
+            raise ValueError("Multi-worker processing not supported for this dataset yet, please use 'num_workers=1'.")
+        return iter(self.iterable_dataset)
+
+    def __len__(self):
+        return self.max_samples_per_device
+
+    def _subsample(self, device_offset, text_batch):
+        # Only return the i-th item out of N sequential items
+        for k, v in text_batch.items():
+            text_batch[k] = v[device_offset:device_offset + 1]
+        return text_batch
+
+    def _shard_dataset(self, dataset):
+        # Verify # of shards
+        filepaths = dataset._ex_iterable.kwargs['filepaths']
+        if self.num_shards != len(filepaths):
+            raise ValueError(f"Found {len(filepaths)} shards, expected {self.num_shards}")
+
+        # Determine how to allocate devices to shards
+        devices_per_shard = 1
+        if self.num_shards < self.world_size:
+            log.warning(
+                f"Not enough unique shards ({self.num_shards}) for world size ({self.world_size}). Splitting shards among devices."
+            )
+            if self.world_size % self.num_shards != 0:
+                raise ValueError(f"Cannot evenly split {self.num_shards} shards among {self.world_size} devices")
+            devices_per_shard = self.world_size // self.num_shards
+        elif self.num_shards % self.world_size != 0:
```

Comment:
since `num_shards` is defined within the functioni (e.g. `1024`), does that mean that using the streaming dataset with 3 or 6 or 7 GPUs is not supported?

Reply:
Sadly, yes... in general it gets pretty tricky to handle non-divisible number of shards, I can think of two options:
* truncate / duplicate some shards to make `new_num_shards` divisible
* figure out some way to iterate through one set of shards normally, and then a second set of shards (the leftovers) via subsampling... but this seems a bit complicated.

For context, the `webdataset` package also just splits the shard list among all CPU workers and assumes that the `num_shards` is divisible.

I think these limitations can be overcome, but I wonder if this is the best place for it, or if we can follow up with some more general adapter that can correct both HF and Webdataset IterableDatasets.

Reply:
Ok with not addressing, just wanted clarification.